packages/vats/src/core/chain-behaviors.js

```diff
@@ -282,6 +282,76 @@ export const makeBridgeManager = async ({
 };
 harden(makeBridgeManager);
 
+// TODO: Refine typing... maybe something like PromiseSpace or PromiseSpaceOf?
+/**
+ * @param { BootstrapPowers } powers
+ */
+export const makeChainStorage = async ({
+  consume: { bridgeManager: bridgeManagerP },
+  produce: { chainStorage: chainStorageP },
+}) => {
+  const bridgeManager = await bridgeManagerP;
+  if (!bridgeManager) {
+    console.warn('Cannot support chainStorage without an actual chain.');
+    chainStorageP.resolve(undefined);
+    return;
+  }
+  const toStorage = message =>
+    E(bridgeManager).toBridge(BRIDGE_ID.STORAGE, message);
+
+  // TODO: Formalize root key.
+  // Must not be any of {activityhash,beansOwing,egress,mailbox},
+  // and must be reserved in sites that use those keys (both Go and JS).
+  const ROOT_KEY = 'published';
+  // TODO: Formalize segment constraints.
+  // Must be nonempty and disallow (unescaped) `.`, and for simplicity
+  // (and future possibility of e.g. escaping) we currently limit to
+  // ASCII alphanumeric plus underscore.
+  const pathSegmentPattern = /^[a-zA-Z0-9_]{1,100}$/;
```

Comment:
A key is a dot-separated list of segments (e.g., "published.foo.bar"). It seems necessary to disallow (unescaped) "." in a segment, and this PR imposes further restrictions:
* Each character of a segment must be an ASCII letter, number, or underscore.
* A segment may not be empty.
* A segment may not be longer than 100 characters (where the differences between character/code point/code unit/etc. can be ignored because of the first restriction).

It does _not_ currently impose a maximum length on the full key, but could easily do so.

This leaves the door open for future extension, in particular to something like backslash escape sequences and/or a larger subset of Unicode (with the requisite care for control characters/homoglyphs/right-to-left subsequences/etc.).

Reply:
I'm slightly in favor of using a slash as separator, rather than dot. What do other cosmos-sdk RPC query paths use? We should probably match their style.

I'm personally a fan of hyphen over underscore, for human-readable things (which don't need to follow e.g. JS identifier syntax), so I'd lobby to allow hyphens too. I'm quite happy with the other limits, but we should probably impose some sort of overall limit to prevent abuse/attack via multi-megabyte keys.